### PR TITLE
Pull of facade elements optimised + extraction of profile and material from mullions implemented

### LINKED ISSUE #1389: Optimize Facade Element Pulls
#### Description:
Currently, pulling facade elements take quite a long time, especially when dealing with large collections of openings, for example. We should look to optimize the frame edge pulling process to minimize performance impact, skipping frame edge profile geometry more effectively and handling the overall conversion of revit window/curtain wall panel/etc to Facade Element more efficiently overall.

#### Steps to reproduce:
Pull many facade elements, and wait.....

#### Expected behaviour:
Faster pulling of elements to improve UX.

#### Test file(s):
[Test File](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/03_Alpha/BHoM/Revit_Toolkit/Pull/PullFacadeElements.gh?csf=1&web=1&e=3Tl90e)

### PR DESCRIPTION
**Pull request by pawelbaran**
### Issues addressed by this PR

Closes #1389

### Test files

Standard test file for pull of facade elements and standard test file for pull of framing elements - **please be very careful while testing** because this PR is touching a few pretty core methods.

[Revit Test Model](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/03_Alpha/BHoM/Revit_Toolkit/Pull/PullFacadeElements_R20.rvt?csf=1&web=1&e=RwbFiS)

[GH Test File](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/03_Alpha/BHoM/Revit_Toolkit/Pull/PullFacadeElements.gh?csf=1&web=1&e=7lF9Km)

### Changelog

- performance of pull of facade elements optimised
- extraction of profile and material from mullions implemented

### Additional comments

Execution time of facade converts has been halved more or less, which cannot be fully experienced due to the to the adapter overheads.

### CODE REVIEW

**Comment by pawelbaran** [comment]
@BHoMBot check core
@BHoMBot check compliance

File: Revit_Core_Engine/Convert/Facade/FromRevit/CurtainWall.cs [hunk 1]
```diff
@@ -95,6 +89,47 @@ public static CurtainWall CurtainWallFromRevit(this Wall wall, RevitSettings set
...
+            if (curtainPanels == null || !curtainPanels.Any())
+                BH.Engine.Base.Compute.RecordError(String.Format("Processing of panels of Revit curtain wall failed. BHoM curtain wall without location has been returned. Revit ElementId: {0}", system.Id.IntegerValue));
+
+            // Get external edges of whole curtain wall
+            List<FrameEdge> allEdges = curtainPanels.SelectMany(x => x.Edges).ToList();
+            List<FrameEdge> extEdges = allEdges.Distinct().Where(x => allEdges.Count(y => x == y) == 1).ToList();
+
+            bHoMCurtainWall = new CurtainWall { ExternalEdges = extEdges, Openings = curtainPanels.ToList(), Name = system.CurtainSystemType.Name };
+
+            bHoMCurtainWall.Name = system.FamilyTypeFullName();
```
Comment (vietle-bh) [inline Revit_Core_Engine/Convert/Facade/FromRevit/CurtainWall.cs:120]: Why doesn't this work?

```suggestion
            bHoMCurtainWall = new CurtainWall { ExternalEdges = extEdges, Openings = curtainPanels.ToList(), Name = system.FamilyTypeFullName() };
```

File: Revit_Core_Engine/Convert/Structure/FromRevit/Profile.cs [hunk 2]
```diff
@@ -131,6 +131,84 @@ public static IProfile ProfileFromRevit(this FamilySymbol familySymbol, RevitSet
...
+
+            BH.oM.Geometry.Line line = representativeMullion.MullionLine();
+            BH.oM.Geometry.Vector direction = line.Direction();
+            foreach (Mullion candidateMullion in allMullionsInWall)
+            {
+                BH.oM.Geometry.Line candidateLine = candidateMullion.MullionLine();
+                if (candidateLine == null || !candidateLine.IsCollinear(line))
+                    continue;
+
+                if ((candidateLine.Start - line.Start).DotProduct(direction) > 0)
```
Comment (vietle-bh) [inline Revit_Core_Engine/Convert/Structure/FromRevit/Profile.cs:182]: What does it mean here when the dot product is > 0?

**Review by vietle-bh — COMMENTED** [review]
Works on first try 👍 

image

File: Revit_Core_Engine/Convert/Structure/FromRevit/Profile.cs [hunk 3]
```diff
@@ -131,6 +131,84 @@ public static IProfile ProfileFromRevit(this FamilySymbol familySymbol, RevitSet
...
+        public static IProfile ProfileFromRevit(this MullionType mullionType, RevitSettings settings = null, Dictionary<string, List<IBHoMObject>> refObjects = null)
+        {
+            settings = settings.DefaultIfNull();
+
+            IProfile profile = refObjects.GetValue<IProfile>(mullionType.Id);
+            if (profile != null)
+                return profile;
+
+            // The algorithm below is so convoluted because mullion type does not have a profile
+            // Instead, one needs to extract the geometry of one of the instances, and it needs to be the one that does not join with its start face
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Revit_Core_Engine/Convert/Structure/FromRevit/Profile.cs:150]: Sorry I was typing while being deep down in code. Tweaked the comment to make it more meaningful

File: Revit_Core_Engine/Convert/Structure/FromRevit/Profile.cs (hunk 2 above)
Comment (pawelbaran) [reply Revit_Core_Engine/Convert/Structure/FromRevit/Profile.cs:190]: I added a comment in the code 👍

**Review by vietle-bh — DISMISSED** [review]
Worked in my tests 👍 

PullFacadeElements:

image

PullFramingLocation:

image

PullBarLocation-Framing:

image

FromLink_PullFramingLocation:

image

FromLink_PullBarLocation-Framing:

image

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Comment by pawelbaran** [comment]
@BHoMBot check installer

**Comment by pawelbaran** [comment]
@BHoMBot check versioning

**Comment by pawelbaran** [comment]
@FraserGreenroyd please do not merge this one - still missing required versioning, I will add it later today

**Comment by FraserGreenroyd** [comment]
> @FraserGreenroyd please do not merge this one - still missing required versioning, I will add it later today

Understood 👍

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Comment by pawelbaran** [comment]
@FraserGreenroyd I have updated versioning, (hopefully) it covers the changes made in this PR. Would be great to get this PR merged rather sooner than later, so if the versioning checks is expected to malfunction for a longer period then I will appreciate if you could have a manual look at the changes vs versioning and skip the check if happy. Thanks!

**Review by vietle-bh — APPROVED** [review]
Reapproving after versioning fix.

**Comment by pawelbaran** [comment]
@FraserGreenroyd this one should be ready to merge I believe

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance

**Pull request merged**